In [20]:
import pandas as pd

df=pd.read_csv("C:/Users/vbopp/Downloads/realest.csv")
df.head(3)

Price  Bedroom  Space  Room   Lot     Tax  Bathroom  Garage  Condition
0   53.0      2.0  967.0   5.0  39.0   652.0       1.5     0.0        0.0
1   55.0      2.0  815.0   5.0  33.0  1000.0       1.0     2.0        1.0
2   56.0      3.0  900.0   5.0  35.0   897.0       1.5     1.0        0.0

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Price      156 non-null    float64
 1   Bedroom    156 non-null    float64
 2   Space      146 non-null    float64
 3   Room       156 non-null    float64
 4   Lot        146 non-null    float64
 5   Tax        147 non-null    float64
 6   Bathroom   156 non-null    float64
 7   Garage     156 non-null    float64
 8   Condition  156 non-null    float64
dtypes: float64(9)
memory usage: 11.2 KB


In [22]:
#
data=df

avg_space=round(data['Space'].mean())
avg_lot=round(data['Lot'].mean())
avg_tax=round(data['Tax'].mean())

data['Space'].fillna(avg_space,inplace=True)
data['Lot'].fillna(avg_lot,inplace=True)
data['Tax'].fillna(avg_tax, inplace=True)

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Price      156 non-null    float64
 1   Bedroom    156 non-null    float64
 2   Space      157 non-null    float64
 3   Room       156 non-null    float64
 4   Lot        157 non-null    float64
 5   Tax        157 non-null    float64
 6   Bathroom   156 non-null    float64
 7   Garage     156 non-null    float64
 8   Condition  156 non-null    float64
dtypes: float64(9)
memory usage: 11.2 KB


In [23]:
data.tail()

data.drop(index=156, inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Price      156 non-null    float64
 1   Bedroom    156 non-null    float64
 2   Space      156 non-null    float64
 3   Room       156 non-null    float64
 4   Lot        156 non-null    float64
 5   Tax        156 non-null    float64
 6   Bathroom   156 non-null    float64
 7   Garage     156 non-null    float64
 8   Condition  156 non-null    float64
dtypes: float64(9)
memory usage: 11.1 KB


In [24]:
#Building a linear regression Model

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Separating features and target variable
X = data.drop(columns=['Price'])  
y = data['Price']

# Spliting data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# training the model
model = LinearRegression()
model.fit(X_train, y_train)

# Predicting on test set
y_pred = model.predict(X_test)

# Evaluating the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')


Mean Squared Error: 65.66713205550892
R-squared: 0.6538551179337232


In [25]:
# Calculate the median price
median_price = data['Price'].median()

# Create a binary target variable where 1 = above/equal to median, 0 = below median
data['price_category'] = data['Price'].apply(lambda x: 1 if x >= median_price else 0)

# Ensure that 'price' column is dropped, and we are only using 'price_category' as the target
a = data.drop(columns=['Price', 'price_category'])
b = data['price_category']

In [28]:
#building a Random forest Classification model

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report


a_train, a_test, b_train, b_test = train_test_split(a, b, test_size=0.2, random_state=42)

# training the model
model = RandomForestClassifier(random_state=42)
model.fit(a_train, b_train)

# Predicting and evaluating the results
b_pred = model.predict(a_test)
accuracy = accuracy_score(b_test, b_pred)
report = classification_report(b_test, b_pred)

print(f'Accuracy: {accuracy}')
print('Classification Report:')
print(report)



Accuracy: 0.90625
Classification Report:
              precision    recall  f1-score   support

           0       0.81      1.00      0.90        13
           1       1.00      0.84      0.91        19

    accuracy                           0.91        32
   macro avg       0.91      0.92      0.91        32
weighted avg       0.92      0.91      0.91        32



The comparison between the linear regression and random forest classification models reveals distinct strengths tailored to different objectives. The linear regression model predicts house prices with a Mean Squared Error (MSE) of 65.67 and an R-squared value of 0.65, indicating that it explains about 65% of the variance in the dataset. This model is beneficial for understanding specific price dynamics but may not capture all influencing factors.

In contrast, the random forest classifier demonstrates impressive performance with an accuracy of 90.63%. It excels at categorizing house prices into two classes based on their relation to the median price, achieving high precision and recall rates, particularly for predicting prices above the median. While it shows strong results in classifying prices, it is slightly less effective for those below the median. Ultimately, the choice between these models depends on whether the goal is to predict exact prices or categorize them into classes. Using both models could provide a comprehensive analysis of the housing market.